### DISCUSSION TITLE
suggestion only for color saturation adjustment

### DISCUSSION THREAD

**Discussion by akores** [mapsforge/mapsforge#1450] [Q&A] [answered]
... in map.rendertheme.XmlUtils.java ... only as suggestion, 
... in map.android.util.AndroidUtil.java ... only as suggestion, 
because for "official mapsforge" should be moved to proper places

this printscreen is with low saturation for original InternalRenderTheme.OSMARENDER.

![saturation-015](https://github.com/mapsforge/mapsforge/assets/92294535/8b6b875a-cd15-4474-98d6-6c061e4cec50)

Color class is different for awt (Color.HSBtoRGB) and android (Color.HSVToColor) , so some additional cleanup and reorganisation , where to insert this.... I leave this decisions to  "official mapsforge" participants ...

map.android.util.AndroidUtil.java :
    /**
     * maxColorSaturation [ 0.0 ... 1.0] ( from 0.0 as gray ... 0.15 pastel ... to 1.0 original theme color )
     * TODO: maybe move to Properties.maxColorSaturation  ...
     */

    public static float maxColorSaturation = 0.15f;
    public static int[] changeSaturationAndroid(int red, int green, int blue, float maxSaturation) {
        int[] rgb = new int[3];
        float[] hsv0 = new float[3];
        Color.RGBToHSV(red, green, blue, hsv0);
        hsv0[1] = Float.min(hsv0[1], maxSaturation);
        int icolor = Color.HSVToColor(hsv0);
        rgb[0] = (icolor >> 16) & 0xff;
        rgb[1] = (icolor >> 8) & 0xff;
        rgb[2] = (icolor >> 0) & 0xff;
        return rgb;
    }

map.rendertheme.XmlUtils.java :

    private static int getColor(GraphicFactory graphicFactory, String colorString, int alpha, int rgbStartIndex, ThemeCallback themeCallback, RenderInstruction origin) {
        int red = Integer.parseInt(colorString.substring(rgbStartIndex, rgbStartIndex + 2), 16);
        int green = Integer.parseInt(colorString.substring(rgbStartIndex + 2, rgbStartIndex + 4), 16);
        int blue = Integer.parseInt(colorString.substring(rgbStartIndex + 4, rgbStartIndex + 6), 16);

        int[] rgb = AndroidUtil.changeSaturationAndroid(red, green, blue, AndroidUtil.maxColorSaturation);

        int color = graphicFactory.createColor(alpha, rgb[0], rgb[1], rgb[2]);
        if (darkTheme) {
            color = GraphicUtils.filterColor(color, Filter.INVERT);
        }
        if (themeCallback != null) {
            color = themeCallback.getColor(origin, color);
        }
        return color;
    }

    
for Awt utils  this should be added and properly organized  (by "official mapsforge" participants)
     
    public static int[] changeSaturationAwt(int red, int green, int blue, float maxSaturation) {
        int[] rgb = new int[3];
        float[] fhsv = new float[3];
        java.awt.Color.RGBtoHSB(red, green, blue, fhsv);
        int icolor = java.awt.Color.HSBtoRGB(fhsv[0], Float.min(fhsv[1], maxSaturation), fhsv[2]);
        rgb[0] = (icolor >> 16) & 0xff;
        rgb[1] = (icolor >> 8) & 0xff;
        rgb[2] = (icolor >> 0) & 0xff;
        return rgb;
    }

**Comment by devemux86** [comment]
There is [ThemeCallback](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/rendertheme/ThemeCallback.java) that can be used to customize the theme colors per render instruction in the application.

**Reply by akores** [reply, marked as answer]
thanks for remind me of [ThemeCallback](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/rendertheme/ThemeCallback.java). my bad, I didn't go into maps-forge deep enough... and also sorry for posting under issues ...

**Comment by akores** [comment]
thanks again for elegant [ThemeCallback](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-map/src/main/java/org/mapsforge/map/rendertheme/ThemeCallback.java) suggestion. so if anyone wants to use such saturation adjustment, here is clean solution for android for lower saturation,( in this example hard coded (for simplicity) max saturation 0.12f ) 
```
   public ThemeCallback pastelColorsCallBack=new ThemeCallback() {
        float[] hsv0 = new float[3];
        @Override
        public int getColor(RenderInstruction origin, int color) {
            Color.colorToHSV(color, hsv0);
            hsv0[1] = Float.min(hsv0[1], 0.12f);
            int newcolor = Color.HSVToColor(hsv0);
            return newcolor;
        }
    };
```
